src/background/sidebar-injector.ts

```diff
@@ -0,0 +1,485 @@
+        } else {
+          // If the content script threw an exception,
+          // frameResults may be null or undefined.
+          //
+          // In that case, fall back to guessing based on the
+          // tab URL
+          return guessContentTypeFromURL(tab.url);
+        }
+      } else {
+        // We cannot inject a content script in order to determine the
+        // file type, so fall back to a URL-based mechanism
+        return guessContentTypeFromURL(tab.url);
+      }
+    }
+
+    /**
+     * Returns true if a tab is displaying a PDF using the PDF.js-based
+     * viewer bundled with the extension.
+     */
+    function isPDFViewerURL(url: string) {
+      return url.startsWith(pdfViewerBaseURL);
+    }
+
+    function isFileURL(url: string) {
+      return url.startsWith('file:');
+    }
+
+    function isSupportedURL(url: string) {
+      // Injection of content scripts is limited to a small number of protocols,
+      // see https://developer.chrome.com/extensions/match_patterns
+      const parsedURL = new URL(url);
+      return ['http:', 'https:', 'ftp:'].includes(parsedURL.protocol);
+    }
+
+    async function injectIntoLocalDocument(tab: Tab) {
+      const type = await detectTabContentType(tab);
+      if (type === CONTENT_TYPE_PDF) {
+        return injectIntoLocalPDF(tab);
+      } else {
+        throw new LocalFileError('Local non-PDF files are not supported');
+      }
+    }
+
+    async function injectIntoRemoteDocument(tab: Tab, config: object) {
+      if (isPDFViewerURL(tab.url)) {
+        return;
+      }
+
+      if (!isSupportedURL(tab.url)) {
+        // Chrome does not permit extensions to inject content scripts
+        // into (chrome*):// URLs and other custom schemes.
+        //
+        // A common case where this happens is when the user has an
+        // extension installed that provides a custom viewer for PDFs
+        // (or some other format). In some cases we could extract the original
+        // URL and open that in the Hypothesis viewer instead.
+        const protocol = tab.url.split(':')[0];
+        throw new RestrictedProtocolError(
+          'Cannot load Hypothesis into ' + protocol + ' pages'
+        );
+      }
+
+      const type = await detectTabContentType(tab);
+
+      if (type === CONTENT_TYPE_PDF) {
+        await injectIntoPDF(tab);
+      } else if (type === CONTENT_TYPE_VITALSOURCE) {
+        await injectIntoVitalSourceReader(tab, config);
+      } else if (type === CONTENT_TYPE_LMS) {
+        // The extension is blocked on LMS assignments to avoid confusion with the
+        // embedded Hypothesis instance. The user can still use the extension on other
+        // pages hosted in the LMS itself.
+        throw new BlockedSiteError(
+          "Hypothesis extension can't be used on Hypothesis LMS assignments"
+        );
+      } else {
+        const result = (await injectIntoHTML(tab, config)) as {
+          installedURL: string;
+        } | null;
```

Comment:
Probably easier to read if the return type is defined in the `injectIntoHTML` function, and in case we need to do some type casting, we do it there, instead of here.

Reply:
The suggestion makes sense. I looked into this though and realized there was a bigger problem - nothing is actually setting `installedURL`! It looks like it used to be part of a copy of the client's boot script that was embedded into this repository - e0bf3fd2a09414170eb991d7837bf6acd821502b.